Read SoA scalars and Eigen columns from a ROOT streamer [12.5.x]

fwyzard opened the pull request:
#### PR description:

Fix the ROOT streamers for the template-based SoA to read back the scalar variables, and to write and read back Eigen objects.

Update the `TestAlpakaAnalyzer` and CUDA `TestPortableAnalyzer` to check that scalars and matrices are read back correctly, and print a more verbose description of the SoA being read.

Update the test scripts so that the unit tests actually fail (exit with a non-zero value) if any of the cmsRun jobs fails.

#### PR validation:

The improved unit tests pass.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport of #39532 to CMSSW_12_5_X for the Alpaka migration.

makortel committed:
Add missing includes

fwyzard committed:
Make the TestAlpakaAnalyzer and CUDA TestPortableAnalyzer more verbose

fwyzard committed:
Make the CUDA and Alpaka write/read tests fail on errors

fwyzard committed:
Implement the ROOT streamer for reading back SoA scalars

fwyzard committed:
Check that scalars are set and read back correctly

Move the printout before the assert, so it can provide more details if
the assertion fails.

fwyzard commented:
backport #39532

fwyzard commented:
type bugfix

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for CMSSW_12_5_X.

It involves the following packages:

- DataFormats/SoATemplate (**heterogeneous**)
- HeterogeneousCore/AlpakaCore (**heterogeneous**)
- HeterogeneousCore/AlpakaTest (**heterogeneous**)


@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@makortel, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ericcano committed:
Add ROOT streaming support for Eigen SoA columns

Store Eigen SoA columns as a strided, linearised array, and implement a
ROOT streamer for reading them back as Eigen objects in SoA columns.

Update the CUDA and Alpaka portable collection tests to fill the Eigen
Matrix column, print part of its content, and check that it is written
and read back correctly.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e38aa6/27837/summary.html
**COMMIT**: 848368c517acd061939380c6945e92120bb8577e
**CMSSW**: CMSSW_12_5_X_2022-09-28-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39533/27837/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testHeterogeneousCoreAlpakaTestWriteRead had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3699454
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3699424
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-09-28-2300+e38aa6/53026/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - Reco comparison had  3 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19868
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_5_X_2022-09-28-2300+e38aa6/53023/triggerResults): no differences found

cmsbuild commented:
Pull request #39533 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
please test

cmsbuild commented:
Pull request #39533 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard committed:
Add missing dependency

cmsbuild commented:
Pull request #39533 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard committed:
Add missing typename specifier

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e38aa6/27857/summary.html
**COMMIT**: b4f48b05c5a835b55e83ba4448a6bd4350fc4219
**CMSSW**: CMSSW_12_5_X_2022-09-29-1100/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39533/27857/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package DataFormats/PortableTestObjects
>> Entering Package DataFormats/SoATemplate
>> Entering Package HeterogeneousCore/AlpakaTest
>> Entering Package HeterogeneousCore/CUDATest
>> Compile sequence completed for CMSSW CMSSW_12_5_X_2022-09-29-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-09-29-1100/tmp/el8_amd64_gcc10/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package CUDADataFormats/PortableTestObjects
Entering library rule at CUDADataFormats/PortableTestObjects
>> Building LCG reflex dict from header file src/CUDADataFormats/PortableTestObjects/src/classes.h
</pre>

cmsbuild commented:
Pull request #39533 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
test parameters:
  - enable_tests: gpu

fwyzard commented:
please test

fwyzard commented:
please test

fwyzard commented:
please test with cmsdist#8108

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e38aa6/27867/summary.html
**COMMIT**: 954edbc5dedbc62d6e1e10acfbaced01d2116b7f
**CMSSW**: CMSSW_12_5_X_2022-09-29-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39533/27867/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3699454
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3699424
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-09-29-2300+e38aa6/53056/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - Reco comparison had  3 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19868
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_5_X_2022-09-29-2300+e38aa6/53052/triggerResults): no differences found

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_12_5_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_12_6_X is complete. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
- This will likely trigger a unit test error, to be cured by merging https://github.com/cms-sw/cmsdist/pull/8108, see https://github.com/cms-sw/cmssw/pull/39532#issuecomment-1264567603
